Feature/A-06

vdvhk12 opened the pull request:
<!--
  템플릿은 아직 PR 작성이 익숙하지 않으신 분들을 위해서 제공하는 가이드입니다!
  리뷰어 또는 이 PR을 보게 될 다른 사람들이 이 PR을 보는데 참고할 수 있는 내용이 있다면 포함해서 작성해주시면 됩니다.
-->

## 📌 과제 설명 <!-- 어떤 걸 만들었는지 대략적으로 설명해주세요 -->
- JwtAuthorizationFilter 구현 close #15 

## 👩‍💻 요구 사항과 구현 내용 <!-- 기능을 Commit 별로 잘개 쪼개고, Commit 별로 설명해주세요 -->
- feat: SECRET_KEY 접근제어자 public으로 수정
- feat: 토큰 유효성 검증 에러코드 추가
- feat: CustomUserDetails 구현
- feat: JwtAuthorizationFilter 구현 & SecurityConfig에 addFilterBefore()로 추가
- 각 커밋 별 설명은 간단하게 커밋 메세지에 있습니다.

## ✅ 피드백 반영사항  <!-- 지난 코드리뷰에서 고친 사항을 적어주세요. 재PR 시에만 사용해 주세요! (재PR 아닌 경우 삭제) -->

## ✅ PR 포인트 & 궁금한 점 <!-- 리뷰어 분들이 집중적으로 보셨으면 하는 내용을 적어주세요 -->

kdo1999 committed:
feat: PasswordEncoder 빈 등록 추가

kdo1999 committed:
feat: MemberRepository.existsByUsername 추가

kdo1999 committed:
comment: 주석 추가

kdo1999 committed:
feat: MemberRepository.existsByNickname 추가

kdo1999 committed:
refactor: Address 오타 수정

kdo1999 committed:
feat: MemberException 추가

kdo1999 committed:
feat: MemberService.signup 추가

kdo1999 committed:
feat: 회원 Username 검증기 추가

kdo1999 committed:
feat: 회원 Password 검증기 추가

kdo1999 committed:
feat: 회원 Nickname 검증기 추가

kdo1999 committed:
feat: Enum 검증기 추가

kdo1999 committed:
feat: 비밀번호, 비밀번호 재확인 검증기 추가

kdo1999 committed:
feat: 검증 순서 정의한 인터페이스 추가

kdo1999 committed:
refactor: default 그룹 설정

kdo1999 committed:
refactor: default groups 수정

kdo1999 committed:
feat: Validation 예외 처리 핸들러 추가

kdo1999 committed:
feat: MemberRepository.findByUsername 추가

kdo1999 committed:
feat: Member 빌더 수정

kdo1999 committed:
test: MemberServiceTest.signup_success 추가

kdo1999 committed:
test: MemberServiceTest.signup_exists_username_fail 추가

kdo1999 committed:
test: MemberServiceTest.signup_exists_nickname_fail 추가

kdo1999 committed:
refactor: 미사용 import 제거

kdo1999 committed:
test: MemberRepositoryTest.exists_username_success 추가

kdo1999 committed:
test: MemberRepositoryTest.exists_nickname_success 추가

kdo1999 committed:
feat: MemberSignupRequest 추가

kdo1999 committed:
chore: Lombok 테스트에서 사용하게 변경

kdo1999 committed:
test: MemberControllerTest.signup_success 추가

kdo1999 committed:
test: MemberControllerTest.signup_username_valid_username_fail 추가

kdo1999 committed:
test: MemberControllerTest.signup_nickname_valid_nickname_fail 추가

kdo1999 committed:
test: MemberControllerTest.signup_password_valid_password_fail

kdo1999 committed:
test: MemberControllerTest.signup_password_match_fail 추가

kdo1999 committed:
test: MemberControllerTest.signup_city_not_blank_fail 추가

kdo1999 committed:
refactor: 유효성 검사 메세지 변경

kdo1999 committed:
test: MemberControllerTest.signup_detail_not_blank_fail 추가

kdo1999 committed:
test: MemberControllerTest.signup_detail_not_blank_fail 수정

kdo1999 committed:
test: MemberControllerTest.signup_country_not_blank_fail 추가

kdo1999 committed:
test: MemberControllerTest.signup_district_not_blank_fail 추가

kdo1999 committed:
test: MemberControllerTest.signup_verifyCode_not_blank_fail 추가

kdo1999 committed:
feat: MemberController 추가

kdo1999 committed:
feat: MemberException 예외 처리 핸들러 추가

kdo1999 committed:
test: MemberControllerTest.signup_nickname_exists_fail 추가

kdo1999 committed:
test: MemberControllerTest.signup_username_exists_fail 추가

kdo1999 committed:
refactor: merge 충돌 부분 수정

vdvhk12 committed:
feat: SECRET_KEY 접근제어자 public으로 수정

- 일단 시크릿 키를 JwtAuthorizationFilter에서 사용하기 위해 public으로 수정
- 추후에 수정할 필요가 있을 것 같음.

vdvhk12 committed:
feat: 토큰 유효성 검증 에러코드 추가

- TOKEN_NOT_VALID: 토큰이 유효하지 않을 떼 401 상태코드
- TOKEN_MISSING: 토큰이 없을 때 401 상태코드

vdvhk12 committed:
feat: CustomUserDetails 구현

- UserDetails와 UserDetailsService의 커스텀 구현체
- CustomUserDetails 필드에 Member를 넣어둠.
  - 멤버 조회 같은 경우에 다시 조회할 필요가 없음.

vdvhk12 committed:
feat: JwtAuthorizationFilter 추가

- 스프링 시큐리티에 addFilterBefore 메서드에 JwtAuthorizationFilter 추가
- UsernamePasswordAuthenticationFilter 전에 JwtAuthorizationFilter가 먼저 동작함.

kdo1999 commented on backend/src/main/java/com/example/backend/global/auth/jwt/JwtAuthorizationFilter.java:
```diff
@@ -0,0 +1,109 @@
+        String token = getTokenFromRequest(request);
+
+        if (token == null) {
+            createErrorInfo(AuthErrorCode.TOKEN_MISSING, request, response);
```
이 부분이랑 아래에 필터 단에서 예외를 처리하지 못해서 ErrorCode로 대체해서 사용한건가요??

vdvhk12 replied:
HttpStatus.UNAUTHORIZED를 사용하지 않고 AuthErroCode를 사용하신 의도를 물어보신걸까요?

kdo1999 replied:
AuthException을 발생시켜서 잡는게 아닌 ErrorCode를 사용하신 의도 입니당

vdvhk12 replied:
필터에서 throw new AuthException으로 처리하면 서버 로그로만 찍히더라구요.
필터가 DispatcherServlet 보다 먼저 동작하니까 예외 핸들러가 AuthException을 못잡는거 같아서,
ErrorCode만 사용해서 바로 클라이언트 쪽으로 response 해줬습니다.

kdo1999 commented on backend/src/main/java/com/example/backend/global/auth/jwt/JwtAuthorizationFilter.java:
```diff
@@ -0,0 +1,109 @@
+            return;
+        } else {
+            CustomUserDetails customUserDetails = (CustomUserDetails) customUserDetailsService.loadUserByUsername(
+                getUsernameFromToken(token));
```
여기에서 발생할 예외는 처리되지 않은 거 같은데 확인 부탁드립니당

vdvhk12 replied:
넵 지금 확인해보겠습니다

vdvhk12 replied:
여기서 어떤 예외가 발생할 수 있을까요? 전 생각나는게 없어서 여쭤봅니다.

kdo1999 replied:
customUserDetailsService.getUsernameFromToken() 에서 회원 조회 실패시 발생하는 예외 입니당

vdvhk12 replied:
회원 조회가 안되면 CustomUserDetailsService에서 UsernameNotFoundException으로 처리 되지 않을까요?

kdo1999 replied:
맞습니다. 
근데 throw로 예외가 발생하면 Filter로 넘어올텐데 이거에대한 예외 처리는 없는거 같아서 말씀드렸습니다.

vdvhk12 replied:
아 이해했습니다. 지금 추가해놓고 다시 push 할게요!

kdo1999 reviewed (commented):
어제 늦게까지 하시느라고 고생하셨습니다.

vdvhk12 committed:
refactor: RedisTemplate 빈 충돌 문제 해결

- RefreshTokenService에서 추가한 의존성에서 RedisConfig 에서 등록한 빈이랑 redis 설정시 기본적으로 등록되는 빈이랑 이름이 같아서 빈 충돌 발생
- RedisConfig에 작성한 빈을 삭제하고, 기본 생성되는 빈을 사용하도록 수정
- @Primany 어노테이션을 붙여도 해결가능한데, redis에서 제공하는 템플릿을 사용

vdvhk12 committed:
refactor: 토큰 유효성 검사, getUsername 메서드 JwtProvider로 이동

- 시크릿 키를 상수로 관리하면 안될 것 같아서, JwtProvider 에서만 처리하도록 함.

vdvhk12 committed:
refactor: 토큰 유효성 검증 패스후에 CustomDetailsService에서 발생한 예외 처리 추가

- 아마 로그인한 상태에서 회원탈퇴를 했을 때 이 예외가 터질 것이라고 예상하는데, USER_NOT_FOUND 로 응답.

vdvhk12 committed:
rebase conflict 수정

vdvhk12 committed:
Merge branch 'main' into feature/A-06
